compiler/circle-eval-diff/src/FileDataLoader.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __FILE_DATA_LOADER_H__
+#define __FILE_DATA_LOADER_H__
+
+#include "DataSet.h"
+
+#include <loco/IR/Node.h>
+
+#include <string>
+#include <luci/IR/CircleNodes.h>
+
+namespace circle_eval_diff
+{
+
+class FileDataLoader
+{
+public:
+  virtual ~FileDataLoader() = default;
+
+public:
+  virtual DataSet load(const std::string &file_path) const = 0;
```

Comment:
@jinevening Good point. I didn't consider that. Actually, I think it seems a bit hard to abstract the case of "one-by-one" because it needs to have some pointer that points at the address of data files or something, which varies from data file format to format. I mean there is no common thing to be interfaced.

Current implementation imports a data file with hdf5 library. Then, is it not that different with my implementation? Because hdf5 library loads the whole data at once. And, mine loads it the same way and store that data into `DataSet` container, which seems just two times more memory usage rather than 1000 times. Or, hdf5 library internally loads the data by chunk not the whole file?

Anyway, I checked the memory usage of both cases.
1. main branch (94c1532e717871fe63864a64ce23ef0305590290)
```bash
$ /usr/bin/time -v ./circle-eval-diff --first_model ~/TRIV2_Model_Zoo/VD/centernet_1216.circle    --second_model ~/TRIV2_Model_Zoo/VD/centernet_1216.circle  --first_input_data ~/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5   --second_input_data ~/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5
Evaluating 0'th data
Evaluating 1'th data
Evaluating 2'th data
Evaluating 3'th data
Evaluating 4'th data
Evaluating 5'th data
Evaluating 6'th data
Evaluating 7'th data
Evaluating 8'th data
Evaluating 9'th data
Evaluation finished. Number of data: 10
Mean Absolute Error (MAE)
MAE for conf is 0
MAE for conf_pool is 0
MAE for boxes is 0

Command being timed: "./circle-eval-diff --first_model /home/seongwoo/TRIV2_Model_Zoo/VD/centernet_1216.circle --second_model /home/seongwoo/TRIV2_Model_Zoo/VD/centernet_1216.circle --first_input_data /home/seongwoo/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5 --second_input_data /home/seongwoo/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5"
        User time (seconds): 344.31
        System time (seconds): 0.11
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:44.49
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 55464
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 25
        Minor (reclaiming a frame) page faults: 86627
        Voluntary context switches: 47
        Involuntary context switches: 1153
        Swaps: 0
        File system inputs: 31456
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```
2. this draft pr
```bash
$ /usr/bin/time -v ./circle-eval-diff --first_model ~/TRIV2_Model_Zoo/VD/centernet_1216.circle    --second_model ~/TRIV2_Model_Zoo/VD/centernet_1216.circle  --first_input_data ~/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5   --second_input_data ~/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5
Evaluating 0'th data
Evaluating 1'th data
Evaluating 2'th data
Evaluating 3'th data
Evaluating 4'th data
Evaluating 5'th data
Evaluating 6'th data
Evaluating 7'th data
Evaluating 8'th data
Evaluating 9'th data
Mean Absolute Error (MAE)
MAE for conf is 0
MAE for conf_pool is 0
MAE for boxes is 0

Command being timed: "./circle-eval-diff --first_model /home/seongwoo/TRIV2_Model_Zoo/VD/centernet_1216.circle --second_model /home/seongwoo/TRIV2_Model_Zoo/VD/centernet_1216.circle --first_input_data /home/seongwoo/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5 --second_input_data /home/seongwoo/TRIV2_Data_Zoo/VD/centernet_1216.input.10.h5"
        User time (seconds): 344.51
        System time (seconds): 0.10
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:44.63
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 71284
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 91044
        Voluntary context switches: 1
        Involuntary context switches: 1019
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```
---
1. Maximum resident set size (kbytes): 55464
2. Maximum resident set size (kbytes): 71284

It seems that the latter is not twice as large as the former.

Reply:
> Because hdf5 library loads the whole data at once.

Is this true? Could you test with 1000 data? And please hide confidential information.

Is it possible to give num_data to user and make the user access data by index? It will enable one-by-one access.

Reply:
> Is this true? Could you test with 1000 data? And please hide confidential information.

Not true. I don't know the detail of hdf5 internal implementation, which is why I tested the real case. I tested again with 1000 data.

1. master
```bash
Command being timed: "./circle-eval-diff --first_model ai_sound.circle --second_model ai_sound.circle --first_input_data ai_sound.input.h5 --second_input_data ai_sound.input.h5
        User time (seconds): 1436.63
        System time (seconds): 0.25
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 24:04.82
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 32104
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 10
        Minor (reclaiming a frame) page faults: 4638
        Voluntary context switches: 27
        Involuntary context switches: 19559
        Swaps: 0
        File system inputs: 64528
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

2. this pr
```bash
Command being timed: "./circle-eval-diff --first_model ai_sound.circle --second_model ai_sound.circle --first_input_data ai_sound.input.h5 --second_input_data ai_sound.input.h5
        User time (seconds): 1435.91
        System time (seconds): 0.35
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 24:04.41
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 94668
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 39
        Minor (reclaiming a frame) page faults: 20049
        Voluntary context switches: 357
        Involuntary context switches: 19453
        Swaps: 0
        File system inputs: 77952
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

---

1. Maximum resident set size (kbytes): 32104
2. Maximum resident set size (kbytes): 94668

It looks like a 3x difference. As I said, it won't be 1000x I think. 

> Is it possible to give num_data to user and make the user access data by index? It will enable one-by-one access.

Umm, on second thought after I said it was impossible to abstract, it could be possible if the derive class has a resource pointer and abstract the method that gives single data rather than the whole data at once as you said. I'll try this approach and test it again:)

Reply:
I've changed the design and result is same with the main branch's :) PTAL.
```bash
Command being timed: "./circle-eval-diff --first_model model.circle --second_model model.circle --first_input_data model.input.h5 --second_input_data model.input.h5"
        User time (seconds): 837.34
        System time (seconds): 0.15
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 13:57.54
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 32916  // NOTE here
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 6
        Minor (reclaiming a frame) page faults: 4647
        Voluntary context switches: 22
        Involuntary context switches: 3097
        Swaps: 0
        File system inputs: 63616
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```